packages/mds-ingest-service/repository/index.ts

```diff
@@ -291,6 +270,71 @@ class IngestReadWriteRepository extends ReadWriteRepository {
     }
   }
 
+  private buildCursor = (cursor: VehicleEventsQueryParams) =>
+    Buffer.from(JSON.stringify(cursor), 'utf-8').toString('base64')
+
+  private parseCursor = (cursor: string): VehicleEventsQueryParams & { limit: number } => {
+    try {
+      return JSON.parse(Buffer.from(cursor, 'base64').toString('utf-8'))
+    } catch (error) {
+      throw new ValidationError('Invalid cursor', error)
+    }
+  }
+
+  /**
+   * TypeORM does not handle joined-models and pagination very well, at all.
+   * The second you supply a .take(N) value, it follows a completely different builder plan
+   * to generate your results in a two-phase query. The results were at least 100x slower
+   * than just doing a join and parsing the objects.
+   *
+   * this method is yanked from the `typeorm-cursor-pagination/Paginator.paginate()` source,
+   * and avoids calling .getMany()
+   */
+  private paginateEventTelemetry = async (
+    query: SelectQueryBuilder<EventEntity>,
+    limit: number,
+    beforeCursor: string | null,
+    afterCursor: string | null,
+    order?: GetVehicleEventsOrderOption
+  ): Promise<PagingResult<EventEntity>> => {
+    const pager = buildPaginator({
+      entity: EventEntity,
+      alias: 'events',
+      paginationKeys: [order?.column ?? 'timestamp', 'id'],
+      query: {
+        limit,
+        order: order?.direction ?? (order?.column === undefined ? 'DESC' : 'ASC'),
+        beforeCursor: beforeCursor ?? undefined, // typeorm-cursor-pagination type weirdness
+        afterCursor: afterCursor ?? undefined // typeorm-cursor-pagination type weirdness
+      }
+    })
+
+    const newQuery = pager['appendPagingQuery'](query) as SelectQueryBuilder<EventEntity>
```

Comment:
Why's the `as` cast needed here?

Reply:
Because I'm sneaking around the compiler in the first place in order to call a private method! I've learned so much after a year of TS.

Reply:
Nevermind, it comes back as `any`, but I pushed the typing to the const rather than have too many `any`'s floating around.